### ISSUE TITLE
Suggestion: Masking & Compositing

### ISSUE THREAD

**Issue by Novarleeir** [mortuusars/Exposure#10]
Since filters are already implemented, it would be neat to use the same slot in the camera to create masks/mattes so that compositing may be done. I'm neither an expert, nor even really all too knowledgeable about how this works so I'm probably getting steps wrong or missing some entirely, but hopefully the general idea gets through.

Setup a scene with a black background:
<img width="650" alt="Setup (Large)" src="https://github.com/mortuusars/Exposure/assets/99646709/e491d95b-493b-4f36-9115-b25136c0e1c3">

Then take a picture with a long exposure time to get a clean mask (I cleaned this up in photoshop because I took a horrible photo, but the point still stands):
<img width="556" alt="maskbell (Large)" src="https://github.com/mortuusars/Exposure/assets/99646709/56030e72-8f91-4d0e-b6e4-2aaaa8dda30b">

Then copy the image onto a pane to craft a plate:
<img width="960" alt="CraftingPlate (Large)" src="https://github.com/mortuusars/Exposure/assets/99646709/a97b85c4-e89a-4775-a9af-e9d12af977b2">

Insert plate into camera:
<img width="405" alt="MaskFilter" src="https://github.com/mortuusars/Exposure/assets/99646709/d4179455-fec1-4d04-9e6d-2851eeb9b989">

and some other steps later... you can do some cool stuff with this such as compositing a train onto another picture:
<img width="541" alt="composite (Large)" src="https://github.com/mortuusars/Exposure/assets/99646709/f816472d-4ac0-4758-9bfc-46982e154e0c">

**Comment by mortuusars** [comment]
Thanks for the suggestion. It's an interesting idea, but I have mixed feelings about this.
Firstly, even though it could be an interesting mechanic, I think it fall a bit outside of the scope of this mod. By doing things like this, we can easily fall into the rabbit hole of ideas and suggestions that stop probably only at the latest digital cameras with 16k video recording and a Photoshop implemented inside of Minecraft. Adding depth to the mod is okay, but it should stay as only the mechanic inside of the game, not separate game on top of the game itself. 
Above text is a bit of an exeggeration, but it should give you some insight on how I think about mod design in general.
The main point is - simplicity is the key.

But even if we ignore the above, there are other things: 
It's not really clear how exactly should it work. Especially in the "some other steps later".
Besides considerable effort required to code this, there are other questions that need to be solved:
How do we get desired mask without some kind of painting system? I don't think it is possible to get a good mask with only shutter speed and different blocks.

As I understand it, compositing in film photography was done on the printing stage, not in camera. For the gameplay purposes it can be made to be in camera, but it's not completely logical. Should it work if we just stick the image in front of the lens? Even though the mod's camera does not have a depth of field to blur the image, it would still be almost black because there would be no light access to in from inside.

That's the first things from the top of my head, but there probably many more details left unresolved about the mechanic.

The lest filter expansion itself is interesting though, i'll note this.

**Comment by Novarleeir** [comment]
You're right, my explanation was lacking. I was referring to the Williams process: https://en.wikipedia.org/wiki/Williams_process
Now obviously that applies to video, but the same works for individual images as well. It involves capturing a subject in front of a black background and repeating exposures until a clean matte has been made.

I actually sort-of achieved this in game by taking a picture (1sec exposure) of an armor stand in front of some black concrete, developing it, then placing it on a wall and taking a picture of the picture, and then once more to get the background blacker and ended up with a nice result. So, with some effort it can be done in-game.
<img width="1241" alt="matte" src="https://github.com/mortuusars/Exposure/assets/99646709/f68f24d1-5ef6-4724-a7d7-53ff64de450c">

I suppose what I was trying to say is that this matte (not the printed version of it but the **_negative_** so that the light being let through is everything _**except**_ the armor stand matte), could be placed into the filter slot and when your desired background picture is taken you get this:
<img width="677" alt="blackmatte" src="https://github.com/mortuusars/Exposure/assets/99646709/1d33b82c-076a-489f-8bc7-4b6b22d42a76">

This is the correct exposure of the armor stand and what will be composited in the final product:
<img width="677" alt="1" src="https://github.com/mortuusars/Exposure/assets/99646709/98bea749-c17a-40e7-8108-1887af94b98f">

Then expose the armor stand onto the unexposed sections of the background image to get something like this:
<img width="677" alt="compositeImage" src="https://github.com/mortuusars/Exposure/assets/99646709/6e5b098a-e5ea-4e1e-99a9-fd565e33b8f5">


What can be done in game currently:
- Creating a matte (albeit a bit difficult to get things lined up and not drifting around)
- A filter slot that alters the image taken

What would have to be implemented:
- Apon placing a matte 'plate' into the camera and taking an image: A pixel on the bufferedImage is blacked out if the corresponding pixel on the matte is below a black value threshold (as it'd be difficult to get perfect blacks in game). Then once all the pixels are compared with the matte it returns the bufferedImage with the required pixels being blacked out.
- And this is the thing you're likely to disagree with as it goes against the simplicity and scope of the mod, which I completely understand: A way to re-expose an already taken negative (or 'combining' two images). Whether it stays in the camera, or a block like the lightroom, this is probably the most crucial thing. The easiest way to handle this for matting, is to simply copy the pixels that are not black over to the second image.

Obviously, this is a fairly niche and overcomplicated process, but this is just an example of something possible with being able to re-expose film. Other applications could include double exposures that leave the ghostly image on top of another like those [80s floating head portraits](https://imgur.com/xYXQWKX), or lightening/darkening images, ect.

Interested to see your thoughts on this.

Edit: Reading this back over I realize that yes, either the white or black sections of the matte would have to be cut out to allow light in. Glass pane + shears? Or perhaps it's not too important and that's getting pedantic for the sake of accuracy?

**Comment by mortuusars** [comment]
The concept is definitely interesting, i'm a bit surprised that the decent mask can be produced by taking picture of a picture. For some reason I didn't think that would work acceptably.

But yeah, you're right, it's still a "fairly niche and overcomplicated process", and it would require a substantial amount of work.

I've actually wanted to make double exposures possible, but quickly discarded that idea because it would require some kind of frame selector control in UI to choose what frame you want to expose, which would be quite a challenge to make it intuitive and not confusing. 

Another good example, I think, is that for the most of development process you could release the shutter even if no film was inserted. It was changed very close to the release. Even though it was realistic (at least I think it is for most film cameras), I decided against it in favor of simplicity and clarity. I had several occasions when I wandered why the camera isn't working properly (when debugging something) and it turned out that no film was inserted, or roll has ended. And I coded this thing, knowing how it works from the start. I can only imagine frustration of the new player when they would realize that shots they have taken were for nothing because no film was inserted. 

And I don't think there is a way to make masking/compositing simple and intuitive. So yeah, that's my thoughts about it. 
Maybe some day, when the mod will have 10's of millions downloads and everybody would be making addons for it like for Create right now, someone will implement this in an addon :)

**Comment by donmor** [comment]
> I've actually wanted to make double exposures possible, but quickly discarded that idea because it would require some kind of frame selector control in UI to choose what frame you want to expose, which would be quite a challenge to make it intuitive and not confusing.

Hey, I have an idea about it (based off film cameras irl, without frame selector stuff):

There can be a key binding for double exposures, regarding the double exposure switch on realistic cameras that have to be pushed and hold during the first exposure. Hold down the "double exposure" key and a "double exposure" mark appear on the bottom bar, then press the shutter. A red "exposed" mark appears after this, indicating the film is exposed but not wound yet (this mark is actually on the counter window irl, and could also blink once right after a normal exposure to be more realistic). You can then either hold the "double exposure" key and press the shutter again to make triple exposure or even more, or just press the shutter to end it up. When it is latter, the "exposed" mark disappears, and the camera is ready for the next photo.

I'm not sure if anyone unfamiliar to photographing can understand it, but anyone who do double exposures irl should be used to it immediately :)

**Comment by mortuusars** [comment]
@donmor I think it could be a good idea to add double exposure to the Lightroom by just printing another photo on an existing photograph (using it as a paper), but I'm struggling to get good results from combining various images in photoshop. 

These are some of the results using screen blending mode (other blend modes look worse):
![image](https://github.com/user-attachments/assets/d69b8f05-8c22-4e6f-9164-549ef6bd457f)
![image](https://github.com/user-attachments/assets/af1dcd60-1395-44e7-9eea-460b1e58bb5a)
![image](https://github.com/user-attachments/assets/b604d298-d32f-4a8d-9bc1-5a1e032facfa)
Also, images above were not dithered (or converted to minecraft palette) after blending, so in reality it will look even worse color-wise.
Maybe I'm not doing it right, but I don't know of a better way of combining exposures.

**Comment by donmor** [comment]
Have you tried adding each two values in R/G/B and then clamping them to 255? This may simulates a irreversible exposure process, just like films irl.

Also, multi-exposures irl uses smaller aperture or higher shutter speed, to reduce light through the lens, to avoid overexposure. Such photos may have conspicuous dark object or region as the first, and then shoot some light object in dark background, as if it is burnt into the first object. Another practice is to have a silhouette of a player, then shoot the milky way with it, to have a "Man of the stars" thing.

**Comment by donmor** [comment]
> Also, multi-exposures irl uses smaller aperture or higher shutter speed, to reduce light through the lens, to avoid overexposure.

So it is necessary to make it intentionally underexposed to have double exposures. On-film compositing should be fine, rather than having 2 underexposed pictures on a film. Expose on a printed photo, however, is not what we do irl.
